kymatio/scattering2d/backend/torch_backend.py

```diff
@@ -0,0 +1,343 @@
+        if not inverse:
+            raise RuntimeError('C2R mode can only be done with an inverse FFT.')
+
+    if not iscomplex(x):
+        raise TypeError('The input should be complex (e.g. last dimension is 2).')
+
+    if not x.is_contiguous():
+        raise RuntimeError('Tensors must be contiguous!')
+
+    if direction == 'C2R':
+        output = torch.irfft(x, 2, normalized=False, onesided=False) * x.shape[-2] * x.shape[-3]
+    elif direction == 'C2C':
+        if inverse:
+            output = torch.ifft(x, 2, normalized=False) * x.shape[-2] * x.shape[-3]
+        else:
+            output = torch.fft(x, 2, normalized=False)
+
+    return output
+
+def cdgmm(A, B, inplace=False):
+    """Complex pointwise multiplication.
+    
+        Complex pointwise multiplication between (batched) tensor A and tensor B.
+
+        Parameters
+        ----------
+        A : tensor
+            A is a complex tensor of size (B, C, M, N, 2).
+        B : tensor
+            B is a complex tensor of size (M, N, 2) or real tensor of (M, N, 1).
+        inplace : boolean, optional
+            If set to True, all the operations are performed inplace.
+
+        Raises
+        ------
+        RuntimeError
+            In the event that the filter B is not a 3-tensor with a last
+            dimension of size 1 or 2, or A and B are not compatible for
+            multiplication.
+        TypeError
+            In the event that A is not complex, or B does not have a final
+            dimension of 1 or 2, or A and B are not of the same dtype, or if
+            A and B are not on the same device.
+
+        Returns
+        -------
+        C : tensor
+            Output tensor of size (B, C, M, N, 2) such that:
+            C[b, c, m, n, :] = A[b, c, m, n, :] * B[m, n, :].
+
+    """
+    if not iscomplex(A):
+        raise TypeError('The input must be complex, indicated by a last '
+                        'dimension of size 2.')
+
+    if B.ndimension() != 3:
+        raise RuntimeError('The filter must be a 3-tensor, with a last '
+                           'dimension of size 1 or 2 to indicate it is real '
+                           'or complex, respectively.')
+
+    if not iscomplex(B) and not isreal(B):
+        raise TypeError('The filter must be complex or real, indicated by a '
+                        'last dimension of size 2 or 1, respectively.')
+
+    if A.shape[-3:-1] != B.shape[-3:-1]:
+        raise RuntimeError('The filters are not compatible for multiplication!')
+
+    if A.dtype is not B.dtype:
+        raise TypeError('A and B must be of the same dtype.')
+
+    if B.device.type == 'cuda':
+        if A.device.type == 'cuda':
+            if A.device.index != B.device.index:
+                raise TypeError('A and B must be on the same GPU!')
+        else:
+            raise TypeError('A must be on GPU.')
+
+    if B.device.type == 'cpu':
+        if A.device.type == 'cuda':
+            raise TypeError('A must be on CPU.')
```

Comment:
Is this good?

Reply:
@janden @eickenberg

Reply:
What should I put down for A instead?

Reply:
“Image”, “Array”, “Input” (like you have above, with “input” and “filter”), or something like that.